### ISSUE TITLE
bstables: columns button functionality

### ISSUE THREAD

**Issue by daN4cat** [jekkos/opensourcepos#563]
The dropdown columns button shows what could be shown or hidden.

Not sure if having unnamed checkboxes makes sense.
Also it would be nice to have a means to store layout configuration by user, so people can get always what they set. This will be handy if let say in the future we want to add more columns and people might want one or another but not all.

![image](https://cloud.githubusercontent.com/assets/636048/15268152/51bc9074-19cd-11e6-8f84-6b59efc9a073.png)

**Comment by jekkos** [comment]
It seems to be all or nothing for this feature. Will think of some kind of jquery rule to modify the DOM and make these empty columns disappear (and thus they'll be always visible)

**Comment by daN4cat** [comment]
I guess it's a tricky one because you could 1, 2, 3 ... maybe more in the future. Maybe hide if no title?

**Comment by daN4cat** [comment]
BTW, do you think persistency of layout config would be possible?

**Comment by jekkos** [comment]
Yes something along those lines, hide if no title, alas there's no easy way to write code to remove it from the DOM. 

For persistency I guess we could do that using localStorage API.. once column visbility is changed then persist the state as an object to this localStorage object.

**Comment by daN4cat** [comment]
Sounds good, but persistency must be by user, right?

**Comment by jekkos** [comment]
yes one could store it per session as well using sessionStorage. I think localStorage is browser wide.

**Comment by jekkos** [comment]
Ok I did a first fix for the empty columns. I didn't see that there was an attribute foreseen to show/hide column names in this dropdown. This is now properly implemented. 

This morning I've added some code to handle the persistence of this selection in session Storage. Code almost done but still facing some issue that sessionStorage only save text (and thus no direct js objects). Will try to fix that later today.

**Comment by daN4cat** [comment]
Maybe you need to serialise in json format?

**Comment by jekkos** [comment]
Yes think that will do the trick. 

Also you suggested to save preferences by user. So I guess I might want to create an entry per user id in localStorage so that the data is persisted after session is gone.

**Comment by daN4cat** [comment]
Yes, I believe that needs to be stored by user (employee) so extra field story.

**Comment by jekkos** [comment]
Column visibility is now persisted. Have tested and it seems to work after login as well. Preferences are bound to current logged in user.

We might want to review other localStorage usage in the application (like for printer selection) to become user bound as well. A bit of namespacing in the config key names might be good as well. Eg using a structure as follows

user_id.module_name.config_option

**Comment by daN4cat** [comment]
I'll test the latest addition in the next days. I'm now abroad and having little time for this.

**Comment by jekkos** [comment]
If anyone else finds the time to test this feature, then be my guest. It's in the master branch now.

**Comment by pablos1rvent** [comment]
I have all the time of my day working on this, what do you need to test? ^^

**Comment by RamkrishnaMondal** [comment]
Already tested on two different localhost servers, persistence working even after logout and re-login, until i clear my browser history.

**Comment by daN4cat** [comment]
OK, I found a bug. Here the scenario:

- from Customer, Supplier or Employee: uncheck the id field
- click on the SMS or Edit icons the form will not open as a model dialog but as a new view
- go back and try again, the form will open as modal dialog
- check the id field
- same sequence as above

So it happens only once when you check/uncheck the field.

**Comment by jekkos** [comment]
This probably happens because the table is rebuilt without server side refresh. Dialog init was done in the loadSuccess callback (so after data was fetched from the server). Think I might want to move dialog initialization to the onPostBody callback instead to handle this kind of situation.

**Comment by jekkos** [comment]
Ok did a fix for this.

**Comment by daN4cat** [comment]
Tested and confirm it's fixed.
